platform/linux-generic/odp_ipsec_sad.c

```diff
@@ -916,32 +946,82 @@ int _odp_ipsec_sa_replay_precheck(ipsec_sa_t *ipsec_sa, uint32_t seq,
 				  odp_ipsec_op_status_t *status)
 {
 	/* Try to be as quick as possible, we will discard packets later */
-	if (ipsec_sa->antireplay &&
-	    seq + IPSEC_ANTIREPLAY_WS <=
-	    (odp_atomic_load_u64(&ipsec_sa->hot.in.antireplay) & 0xffffffff)) {
+	if (ipsec_sa->antireplay && ((seq + ipsec_sa->hot.in.win_size) <=
+	    (odp_atomic_load_u64(&ipsec_sa->hot.in.wintop_seq) & 0xffffffff))) {
 		status->error.antireplay = 1;
 		return -1;
 	}
 
 	return 0;
 }
 
-int _odp_ipsec_sa_replay_update(ipsec_sa_t *ipsec_sa, uint32_t seq,
-				odp_ipsec_op_status_t *status)
+static inline int ipsec_wslarge_replay_update(ipsec_sa_t *ipsec_sa, uint32_t seq,
+					      odp_ipsec_op_status_t *status)
 {
-	int cas = 0;
-	uint64_t state, new_state;
+	uint32_t bucket, wintop_bucket, new_bucket;
+	uint32_t bkt_diff, bkt_cnt;
+	uint64_t bit = 0;
+
+	odp_spinlock_lock(&ipsec_sa->hot.in.lock);
+	if ((seq + ipsec_sa->hot.in.win_size) <=
+	    (odp_atomic_load_u64(&ipsec_sa->hot.in.wintop_seq) & 0xffffffff))
+		goto ar_err;
+
+	bucket = (seq >> IPSEC_AR_WIN_BUCKET_BITS);
+
+	/* Check if the seq is within the range */
+	if (seq > odp_atomic_load_u64(&ipsec_sa->hot.in.wintop_seq)) {
+		wintop_bucket =	odp_atomic_load_u64(&ipsec_sa->hot.in.wintop_seq) >>
+				IPSEC_AR_WIN_BUCKET_BITS;
+		bkt_diff = bucket - wintop_bucket;
+
+		/* Seq is way after the range of AR window size */
+		if (bkt_diff > ipsec_sa->hot.in.nb_bucket)
+			bkt_diff = ipsec_sa->hot.in.nb_bucket;
+
+		for (bkt_cnt = 0; bkt_cnt < bkt_diff; bkt_cnt++) {
+			new_bucket = (bkt_cnt + wintop_bucket + 1) &
+				     (ipsec_sa->hot.in.nb_bucket - 1);
+			ipsec_sa->hot.in.bucket_arr[new_bucket] = 0;
+		}
+
+		/* AR window top sequence number */
+		odp_atomic_store_u64(&ipsec_sa->hot.in.wintop_seq, seq);
```

Comment:
Maybe the logic of this function should be driven even more by bucket index than the sequence number so that wintop_seq (or top_bucket or whatever it would be called), would always correspond to the topmost sequence number of the topmost bucket that is in use. In other words, once a new bucket is taken into use, we could just start using all the bits as if the topmost sequence number was already received.

Currently, if packets are processed in the sequence number order (as is quite likely), every packet would have to update wintop_seq (which may not be cheap due to sharing of the cache line with other cores) and run through the logic associated with wintop increase. So instead of doing that 64 times per bucket, we could do it just once per bucket and effectively increase wintop by 64 when a new bucket is taken into use (but then maybe we could just count buckets instead of sequence numbers also for wintop).